### PR TITLE
Run command

### PR DESCRIPTION
**Pull request by sergey-borovkov**
## What does this PR do?
Implements run command for edge app server emulation.
Also adds an ignore list.

## GitHub issue or Phabricator ticket number?
https://phorge.wireload.net/T7251
## How has this been tested?
tests + manual
## Checklist before merging

- [x] If have added tests.
- [x] Is this a new feature? If yes, please write one phrase about this update.
- [ ] I've attached relevant screenshots (if relevant).

### CODE REVIEW

File: src/cli.rs [hunk 1]
```diff
@@ -174,11 +159,38 @@ pub struct Headers {
...
+    }
+}
+
+impl KeyValuePairs for Secrets {
+    fn new(pairs: Vec<(String, String)>) -> Self {
+        Secrets { secrets: pairs }
+    }
+}
+
+fn parse_key_values<T: KeyValuePairs>(s: &str) -> Result<T, ParseError> {
```
Comment (korvyashka) [inline src/cli.rs:181]: Generic - cool :)

File: src/cli.rs [hunk 2]
```diff
@@ -909,6 +938,27 @@ pub fn handle_cli_edge_app_command(command: &EdgeAppCommands) {
...
+            let secrets = if let Some(secret_pairs) = secrets {
+                secret_pairs.secrets.clone()
+            } else {
+                Vec::new()
+            };
+            let path = match path {
+                Some(path) => PathBuf::from(path),
+                None => env::current_dir().unwrap(),
+            };
+            if !path.join("mock-data.yml").exists() {
```
Comment (korvyashka) [inline src/cli.rs:951]: Used several times over the code - let it be constant.

File: src/commands/edge_app.rs [hunk 3]
```diff
@@ -234,6 +236,38 @@ impl EdgeAppCommand {
...
+    pub fn run(&self, path: &Path, secrets: Vec<(String, String)>) -> Result<(), anyhow::Error> {
+        let address_shared = Arc::new(Mutex::new(None));
+        let address_clone = address_shared.clone();
+
+        let runtime = tokio::runtime::Runtime::new().unwrap();
+        let path = path.to_path_buf();
+        runtime.block_on(async {
+            tokio::spawn(async move {
+                let address = run_server(path.as_path(), secrets).await.unwrap();
+                let mut locked_address = address_clone.lock().unwrap();
```
Comment (korvyashka) [inline src/commands/edge_app.rs:248]: Hm, strange - it is used here and immediately reassigned on next line... Will lock persist after reassignment?

File: src/commands/edge_app.rs [hunk 4]
```diff
@@ -234,6 +236,38 @@ impl EdgeAppCommand {
...
+                tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+            }
+
+            println!(
+                "Edge app emulator is running at {}/index.html",
+                address_shared.lock().unwrap().as_ref().unwrap()
+            );
+
+            loop {
+                tokio::time::sleep(std::time::Duration::from_secs(3600)).await;
```
Comment (korvyashka) [inline src/commands/edge_app.rs:264]: Why just not forever?

File: src/commands/edge_app.rs [hunk 5]
```diff
@@ -234,6 +236,38 @@ impl EdgeAppCommand {
...
+            })
+            .await
+            .unwrap();
+
+            while address_shared.lock().unwrap().is_none() {
+                tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+            }
+
+            println!(
+                "Edge app emulator is running at {}/index.html",
```
Comment (korvyashka) [inline src/commands/edge_app.rs:259]: Hm... index.html?

File: src/commands/edge_app.rs [hunk 6]
```diff
@@ -333,6 +367,37 @@ impl EdgeAppCommand {
...
+        let edge_app_dir = path.parent().ok_or(CommandError::MissingField)?;
+
+        let default_metadata = Metadata::default();
+
+        let mut settings: HashMap<String, serde_yaml::Value> = HashMap::new();
+        for setting in &manifest.settings {
+            if setting.type_ != "secret" {
+                settings.insert(
+                    setting.title.clone(),
+                    serde_yaml::Value::String(setting.default_value.clone()),
```
Comment (korvyashka) [inline src/commands/edge_app.rs:382]: If default_values are used for settings - may be for dev server secrets can also use default_values?

File: src/commands/edge_app_utils.rs [hunk 7]
```diff
@@ -41,18 +42,35 @@ impl FileChanges {
...
+    }
+
+    // Check if the path should be ignored according to .ignore rules
+    let matched = ignore.matched(entry.path(), entry.file_type().is_dir());
+    !matched.is_ignore()
+}
+
+fn load_gitignore(path: &Path) -> Result<Gitignore, CommandError> {
+    let mut builder = GitignoreBuilder::new(path);
+    builder.add(path.join(".ignore"));
```
Comment (korvyashka) [inline src/commands/edge_app_utils.rs:58]: .gitignore?

File: src/commands/edge_app.rs (hunk 4 above)
Comment (sergey-borovkov) [reply src/commands/edge_app.rs:264]: I did not find forever method

File: src/commands/edge_app.rs (hunk 3 above)
Comment (sergey-borovkov) [reply src/commands/edge_app.rs:248]: I am not sure I understand

File: src/commands/edge_app.rs (hunk 5 above)
Comment (sergey-borovkov) [reply src/commands/edge_app.rs:259]: Yes, it is entry point of edge apps... just like in player

File: src/commands/edge_app.rs (hunk 6 above)
Comment (sergey-borovkov) [reply src/commands/edge_app.rs:382]: No, we don't want to store secret values in the file so users don't commit their secrets to git

File: src/commands/edge_app.rs (hunk 3 above)
Comment (korvyashka) [reply src/commands/edge_app.rs:248]: First line is
locked_address = address_clone.....
second is
*locked_address = Some(address)

My question is if first line actually do smth as you imediately replace locked_address value.

File: src/commands/edge_app.rs (hunk 4 above)
Comment (korvyashka) [reply src/commands/edge_app.rs:264]: I see there are some hacks for tokio, though they are not pretty(
https://stackoverflow.com/questions/66721384/how-can-i-make-my-program-sleep-for-a-long-time-while-staying-responsive

File: src/commands/edge_app.rs (hunk 3 above)
Comment (sergey-borovkov) [reply src/commands/edge_app.rs:248]: I still don't understand... it is mutex that calls lock(). I lock first then change the value

File: src/commands/edge_app.rs (hunk 4 above)
Comment (sergey-borovkov) [reply src/commands/edge_app.rs:264]: No point in changing this then I think. Still won't look great.

File: src/commands/edge_app.rs (hunk 3 above)
Comment (korvyashka) [reply src/commands/edge_app.rs:248]: Ok, just was trying to figure out how it works - nvm

**Review by korvyashka — APPROVED** [review]

**Pull request merged**